typing/path.ml

```diff
@@ -88,6 +93,38 @@ let rec scope = function
   | Pdot(p, _) | Pextra_ty (p, _) -> scope p
   | Papply(p1, p2) -> Int.max (scope p1) (scope p2)
 
+let subst id_map p =
+  let changed = ref false in
+  let rec aux = function
+  | Pident id ->
+    begin match List.find (fun (i, _) -> Ident.same i id) id_map with
+      | (_, p) -> changed := true; p
+      | exception Not_found -> Pident id
+    end
+  | Pdot(p, s) -> Pdot(aux p, s)
+  | Pextra_ty(p, e) -> Pextra_ty(aux p, e)
+  | Papply(p1, p2) -> Papply(aux p1, aux p2)
+  in
+  let p' = aux p in
+  if !changed then p' else p
+
+let check_for_unbound_unscoped_idents idl p =
+  let exception Escape of Ident.Unscoped.t in
+  let rec aux = function
+      Pident id ->
+        begin match Ident.find_unscoped id with
+        | None -> ()
+        | Some us ->
+            if Ident.Unscoped.Set.exists (Ident.Unscoped.same us) idl
+            then ()
+            else raise (Escape us)
+        end
+    | Pdot (p, _) | Pextra_ty (p, _) -> aux p
+    | Papply (p1, p2) -> aux p1; aux p2
+  in match aux p with
```

Comment:
I would also propose to add a short path for the `Ident.Unscoped.Set.is_empty idl` case.